tests/rptest/tests/scramful_eos_test.py

```diff
@@ -0,0 +1,170 @@
+#
+# Use of this software is governed by the Business Source License
+# included in the file licenses/BSL.md
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0
+
+from ducktape.mark.resource import cluster
+
+from rptest.clients.types import TopicSpec
+from rptest.tests.redpanda_test import RedpandaTest
+from rptest.clients.rpk import RpkTool
+from time import sleep
+
+from confluent_kafka import (Producer, KafkaException)
+
+
+def on_delivery(err, msg):
+    if err is not None:
+        raise KafkaException(err)
+
+
+class ScramfulEosTest(RedpandaTest):
+    topics = [TopicSpec(name="topic1")]
+
+    def __init__(self, test_context):
+        extra_rp_conf = {
+            "enable_idempotence": True,
+            "enable_transactions": True,
+            "id_allocator_replication": 3,
+            "transaction_coordinator_replication": 3,
+            "id_allocator_replication": 3,
+            "default_topic_replications": 3,
+            "default_topic_partitions": 1,
+            "enable_leader_balancer": False,
+            "enable_auto_rebalance_on_node_add": False,
+            "enable_sasl": True
+        }
+
+        super(ScramfulEosTest, self).__init__(test_context=test_context,
+                                              extra_rp_conf=extra_rp_conf)
+
+    def retry(self, func, retries, accepted_error_code):
+        while True:
+            retries -= 1
+            try:
+                func()
+                break
+            except KafkaException as e:
+                if retries == 0:
+                    raise e
+                assert e.args[0].code(
+                ) == accepted_error_code, f"only {accepted_error_code} error code is expected"
+                sleep(1)
+
+    def write_by_bob(self, algorithm):
+        producer = Producer({
+            "bootstrap.servers": self.redpanda.brokers(),
+            "enable.idempotence": True,
+            "retries": 5,
+            "sasl.mechanism": algorithm,
+            "security.protocol": "SASL_PLAINTEXT",
+            "sasl.username": "bob",
+            "sasl.password": "bob"
+        })
+        producer.produce("topic1",
+                         key="key1".encode('utf-8'),
+                         value="value1".encode('utf-8'),
+                         callback=on_delivery)
+        producer.flush()
+
+    @cluster(num_nodes=3)
+    def test_idempotent_write_fails(self):
+        username, password, algorithm = self.redpanda.SUPERUSER_CREDENTIALS
+
+        rpk = RpkTool(self.redpanda)
+        rpk.sasl_create_user("bob", "bob", algorithm)
+
+        try:
+            self.write_by_bob(algorithm)
+            assert False, "bob should not have access to topic1"
+        except AssertionError as e:
+            raise e
+        except KafkaException as e:
+            assert e.args[0].code() == 29, "auth error is expected"
```

Comment:
nit: you could probably do

```
from kafka import errors as kerr
assert e.args[0].code() == kerr.TopicAuthorizationFailedError.errno
```

See: https://github.com/dpkp/kafka-python/blob/master/kafka/errors.py#L343